### ISSUE TITLE
Could not open JDBC Connection for transaction, already value [org.springframework.jdbc.datasource.ConnectionHolder] for key [io.zonky.test.db.provider.support.BlockingDatabaseWrapper] bound to thread

### ISSUE THREAD

**Issue by cfecherolle** [zonkyio/embedded-database-spring-test#283]
Hi,

I recently migrated my Spring Boot 2 application to Spring Boot 3. With this migration came a lot of dependencies updates, as expected, and there is now a specific use case where my integration tests using the embedded PostgreSQL database won't run.

I get this exception:
```
org.springframework.transaction.CannotCreateTransactionException: Could not open JDBC Connection for transaction

	at org.springframework.data.transaction.ChainedTransactionManager.getTransaction(ChainedTransactionManager.java:142)
	at org.springframework.data.transaction.ChainedTransactionManager.getTransaction(ChainedTransactionManager.java:70)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:601)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:385)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:244)
	at jdk.proxy2/jdk.proxy2.$Proxy230.getLastJobExecution(Unknown Source)
	at org.springframework.batch.core.launch.support.SimpleJobLauncher.run(SimpleJobLauncher.java:111)
	at org.springframework.batch.core.launch.support.TaskExecutorJobLauncher.run(TaskExecutorJobLauncher.java:59)
	at com.okeenea.evelity.product.batch.exportplans.ExportPlacePlansPdfTest.whenRunJob(ExportPlacePlansPdfTest.java:189)
	at com.okeenea.evelity.product.batch.exportplans.ExportPlacePlansPdfTest.testShouldConfigureJobWithCorrectSteps(ExportPlacePlansPdfTest.java:138)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: org.springframework.transaction.CannotCreateTransactionException: Could not open JDBC Connection for transaction
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin(DataSourceTransactionManager.java:313)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.startTransaction(AbstractPlatformTransactionManager.java:400)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:373)
	at org.springframework.data.transaction.MultiTransactionStatus.registerTransactionManager(MultiTransactionStatus.java:69)
	at org.springframework.data.transaction.ChainedTransactionManager.getTransaction(ChainedTransactionManager.java:121)
	... 81 more
Caused by: java.lang.IllegalStateException: Already value [org.springframework.jdbc.datasource.ConnectionHolder@85b4160] for key [io.zonky.test.db.provider.support.BlockingDatabaseWrapper@75c2692b] bound to thread
	at org.springframework.transaction.support.TransactionSynchronizationManager.bindResource(TransactionSynchronizationManager.java:182)
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin(DataSourceTransactionManager.java:304)
	... 85 more
```

I have narrowed the issue to a weird situation: I'm using `@Sql` annotations (two, to be exact) to execute SQL scripts on my database before running the tests.
These annotations trigger a call to `executeSqlScripts` in `SqlScriptsTestExecutionListener` and seem to work as intended.

- When the first script is executed, a call to `bindResource` in `TransactionSynchronizationManager` is made in order to start the transaction.
- This call binds a `BlockingDatabaseWrapper` by adding it to the `resources` map. So far, so good.
- Since I'm using a `ChainedTransactionManager` with two datasources, a second `BlockingDatabaseWrapper` resource is then bound using `bindResource` and added to the map too.
- Then, thanks to `doCleanupAfterCompletion` of `DatasourceTransactionManager`, the first resource is unbound (removed from the map).
- Then, the second resource is unbound too, leaving the map empty after running the SQL scripts, as expected.
- My second `@Sql` script is then executed, and the same steps are repeated. The resources map is left empty, as expected again.
- Then, my integration test starts. It's a Spring Boot test which needs to run a Spring Batch job. It first tries to get the last job instance data from the database before starting, and to do so, binds a `BlockingDatabaseWrapper` resource corresponding to the datasource I'm using with Spring Batch (`bindResource` is called in `doGetConnection` of `DataSourceUtils`)
- The datasource gets bound once.
- Then, `bindResource` is called again in `doBegin` method `DatasourceTransactionManager` and tries to bind the resource that was bound just before, resulting in the error that I copy-pasted above.

I'm not 100% sure this has something to do with the `embedded-database-spring-test`, but the Spring Batch jobs start fine when I launch then in "production" code with an actual database configuration so, I'm wondering whether it's related or not. Feel free to tell me if it seems like it's unrelated though.

Thanks in advance!

**Comment by tomix26** [comment]
Hey @cfecherolle, thanks for the comprehensive report. First, I must say that such problems are very difficult to solve even when one has access to a debugger and can see exactly what's happening in the code, let alone over the internet like this 🙂 

I also can't confirm with 100% certainty whether this problem could be caused by the Zonky library or not. If you're using refresh mode, you need to be aware that between individual tests, the previous database might be discarded and a new one created. This means that the data source instance gets replaced, which results in transactions from the previous test not being propagated to the next test. However, I don't think this is your case.

In the problem description, you mention using `ChainedTransactionManager` with two data sources. What data sources are these? Are they both zonky embedded databases? And if so, are they two distinct data sources, or are you referencing the same data source instance in both `DatasourceTransactionManagers`?

I'm asking because the `DatasourceTransactionManager` implementation holds active transactions in a thread local map that uses the data source instance as a key. So if you have two `DatasourceTransactionManagers` referencing the same data source instance, they actually share this map of active transactions. Therefore, when the first `DatasourceTransactionManager` calls the `bindResource` method, this resource gets registered for both transaction managers, which could cause the error described above.

In such a case, the solution would be to create two separate databases, see https://github.com/zonkyio/embedded-database-spring-test#creating-multiple-databases-within-a-single-test-class, one for each `DatasourceTransactionManager`. But this is just my speculation. It's possible that the problem is caused by something different.

If you discover any new findings, let me know, I'm always happy to help.

**Comment by cfecherolle** [comment]
Hey @tomix26, thank you for taking the time to write such a thorough answer!

I think the `ChainedTransactionManager` was a bit weird to begin with, I originally set it up in my tests to replace Atomikos because it was easier at that time, and I did not need actual JTA handling in my tests.

Now that I have updated a lot of things related to Spring Boot 3 (including Atomikos) it was actually really easy to just switch to the same beans I have for production code and stop using a dummy `ChainedTransactionManager` in my tests.
As I stopped using it, the issue went away.

I guess it was totally unrelated to this library, after all. Thanks again for your kind answer and support!
